src/Parsers/IAST.cpp

```diff
@@ -26,6 +26,39 @@ const char * IAST::hilite_substitution = "\033[1;36m";
 const char * IAST::hilite_none         = "\033[0m";
 
 
+IAST::~IAST()
+{
+    /// If deleter was set, move our children to it.
+    /// Will avoid recursive destruction.
+    if (deleter)
+    {
+        deleter->push_back(std::move(children));
+        return;
+    }
+
+    std::list<ASTs> queue;
+    queue.push_back(std::move(children));
```

Comment:
These push_back's can throw bad_alloc and this might lead to double exception if the dtor itself was called during unwinding. Using something like boost intrusive slist might help to make this safe

Reply:
Yes, and even if we use std::list for children, we can use slice.
I will try to do it.
(For now I think it is not likely to get bad_alloc for small allocation).